extract.go

```diff
@@ -31,12 +31,85 @@ type Extractor struct {
 	Devs      bool
 	Sockets   bool
 	Logger    Logger
+	Ops       FsOps
 	cleanups  []func() error
 }
 
+type FsOps interface {
+	Chmod(string, os.FileMode) error
+	Chown(string, int, int) error
+	Mknod(string, FileInfo) error
+}
+
+type GoFsOps struct{}
+
+func (g GoFsOps) Chmod(name string, mode os.FileMode) error {
+	return os.Chmod(name, mode)
+}
+
+func (g GoFsOps) Chown(name string, uid, gid int) error {
+	return os.Chown(name, uid, gid)
+}
+
+func (g GoFsOps) Mknod(path string, info FileInfo) error {
+	stat := info.Sys().(syscall.Stat_t)
+	return syscall.Mknod(path, DefaultFilePerm, int(stat.Rdev))
+}
+
+// Golang's os.Chown, os.Chmod, syscall.Mknod, make syscalls
+// which are missed by fakeroot's LD_PRELOAD of those filesystem operations.
+// In order to work with fakeroot, we execute the programs
+// 'chown', 'chmod', 'mknod' which are expected to work with fakeroot.
+type FakerootOps struct{}
+
+func (f FakerootOps) Chmod(name string, mode os.FileMode) error {
+	return exec.Command("chmod", fmt.Sprintf("%04o", mode.Perm()), name).Run()
+}
+
+func (f FakerootOps) Chown(name string, uid, gid int) error {
+	return exec.Command("chown", "--no-dereference", fmt.Sprintf("%d:%d", uid, gid), name).Run()
+}
+
+func (f FakerootOps) Mknod(path string, info FileInfo) error {
+	// dtype is b (block), c (char), p (fifo)
+	var dtype string
+	stat := info.Sys().(syscall.Stat_t)
+	majMin := []string{fmt.Sprintf("%d", stat.Rdev/256), fmt.Sprintf("%d", stat.Rdev%256)}
+
+	if info.FMode&os.ModeCharDevice != 0 {
+		dtype = "c"
+	} else if info.FMode&os.ModeDevice != 0 {
+		dtype = "b"
+	} else if info.FMode&os.ModeNamedPipe != 0 {
+		dtype = "p"
+		majMin = []string{}
+	} else {
+		return fmt.Errorf("%s is not a char, block or fifo", path)
+	}
+
+	cmd := append(
+		[]string{"mknod", fmt.Sprintf("--mode=%o", info.FMode.Perm()),
+			path, dtype}, majMin...)
+	if err := exec.Command(cmd[0], cmd[1:]...).Run(); err != nil {
+		if PathExists(path) {
+			return os.ErrExist
```

Comment:
My only thought here is maybe you want to use errors.Wrapf() so you get a stack trace, but it's not significant.

Reply:
Opened issue https://github.com/anuvu/squashfs/issues/13